### PR TITLE
feat: rewrite rounding as a string-digit algorithm; support all RoundingMode modes and bcdivmod (closes #36, #39)

### LINKED ISSUE #36: Add support for additional RoundingMode enums (TowardsZero, AwayFromZero, NegativeInfinity)
## Summary

Currently, the bcmath polyfill only supports 4 of the 7 RoundingMode enum values introduced in PHP 8.4:

**Currently Supported:**
- `RoundingMode::HalfAwayFromZero` (maps to `PHP_ROUND_HALF_UP`)
- `RoundingMode::HalfTowardsZero` (maps to `PHP_ROUND_HALF_DOWN`) 
- `RoundingMode::HalfEven` (maps to `PHP_ROUND_HALF_EVEN`)
- `RoundingMode::HalfOdd` (maps to `PHP_ROUND_HALF_ODD`)

**Not Yet Supported:**
- `RoundingMode::TowardsZero`
- `RoundingMode::AwayFromZero` 
- `RoundingMode::NegativeInfinity`

These three modes currently throw a `ValueError` with the message "RoundingMode::{mode} is not supported" when used with `BCMath::round()`.

## Proposed Implementation

Add support for the remaining three rounding modes in the `BCMath::round()` method:

1. **TowardsZero**: Always round towards zero (truncate)
   - Positive numbers: round down (floor)
   - Negative numbers: round up (ceil)

2. **AwayFromZero**: Always round away from zero
   - Positive numbers: round up (ceil)
   - Negative numbers: round down (floor)

3. **NegativeInfinity**: Always round towards negative infinity (floor)
   - Both positive and negative numbers use floor behavior

## Benefits

- Complete compatibility with PHP 8.4's native RoundingMode enum
- Consistent behavior across all PHP versions (8.1-8.4+)
- Better developer experience when migrating to/from native bcmath extension

## Implementation Notes

The implementation should:
- Update `BCMath::convertRoundingMode()` to handle the new modes
- Update `BCMath::bcroundHelper()` or create specialized logic for these modes
- Add comprehensive test coverage for all rounding scenarios
- Maintain backward compatibility with existing functionality

## Acceptance Criteria

- [ ] `RoundingMode::TowardsZero` works correctly for positive and negative numbers
- [ ] `RoundingMode::AwayFromZero` works correctly for positive and negative numbers  
- [ ] `RoundingMode::NegativeInfinity` works correctly for positive and negative numbers
- [ ] All edge cases are handled (zero, very small numbers, large numbers)
- [ ] Test coverage includes precision scenarios and boundary conditions
- [ ] Documentation is updated to reflect the new supported modes

### LINKED ISSUE #39: Add support for bcdivmod function
## Summary

The `bcdivmod` function is currently not supported in the bcmath polyfill. This function performs division and modulo operations in a single call, returning both the quotient and remainder.

## Details

The `bcdivmod` function was introduced in PHP 8.4 and has the following signature:
```php
bcdivmod(string $num1, string $num2, ?int $scale = null): array
```

It returns an array with two elements:
- `[0]` - The quotient (result of `bcdiv($num1, $num2, $scale)`)
- `[1]` - The remainder (result of `bcmod($num1, $num2, $scale)`)

## Implementation Notes

- Should be implemented in both `lib/bcmath.php` and `src/BCMath.php`
- Need to handle the optional `$scale` parameter correctly
- Should return an array with exactly two string elements
- Error handling should match PHP's native behavior
- Add comprehensive tests to verify correctness

## References

- [PHP Manual: bcdivmod](https://www.php.net/manual/en/function.bcdivmod.php)

### PR DESCRIPTION
**Pull request by nanasess**
## Overview

Following the source-level comparison with `symfony/polyfill-php84` (#81), this PR **rewrites the rounding path from a phpseclib `BigInteger`-based implementation to pure string-digit arithmetic** and fills the remaining feature gaps. It combines the previously-planned PRs A/B/C into one, using a clean re-implementation (no code copied).

- **fix 1**: make `RoundingMode` a **pure enum** (matching native PHP 8.4)
- **fix 2**: rename the bc wrapper parameter `$operand` → `$num` (enables named arguments)
- **closes #36**: implement the four directional modes (`TowardsZero`, `AwayFromZero`, `PositiveInfinity`, `NegativeInfinity`)
- **closes #39**: implement `bcdivmod()` on top of `BCMath::div`/`mod` (works without the native extension)
- **docs**: document `bcdivmod`, all RoundingMode modes, and interoperability with `symfony/polyfill-php84`

## Why rewrite the rounding

The previous `bcround()` fell back to `round((float) ...)` for `HalfEven`/`HalfOdd`, **abandoning arbitrary precision** and returning wrong results for large or high-precision inputs (see #81):

```
bcround('12345678901234567890.5', 0, RoundingMode::HalfEven)
  old pkg : 12345678901234567168   <- float corrupts the low-order digits
  native  : 12345678901234567890
  new     : 12345678901234567890   OK
```

symfony processes every mode with pure string arithmetic (its rounding does **not** call the native bc extension), so adopting that approach improves **correctness and performance at the same time**.

## Changes

| File | Change |
|---|---|
| `lib/RoundingMode.php` | backed string enum → pure enum |
| `lib/bcmath.php` | `$operand` → `$num`; add `bcdivmod` wrapper |
| `src/BCMath.php` | rewrite the rounding core as string-digit arithmetic (all 8 modes + negative precision); rework `convertRoundingMode`; route `floor`/`ceil` through the core; add `divmod()` |
| `tests/BCMathTest.php` | replace the "directional modes throw" expectations with correct-result assertions; add high-precision regression and `bcdivmod` tests |
| `.github/workflows/ci.yml` | remove 8 now-passing phpt tests from `--skip` |
| `phpstan.neon.dist` | adjust ignores for the pure-enum change; ignore a pre-existing (main) PHP 8.1-only `in_array` false positive |
| `README.md` | document new functions; add a three-way comparison with `symfony/polyfill-php84`; note the coexistence behavior |

> The rounding algorithm is a **clean re-implementation** inspired by `symfony/polyfill-php84` (MIT); no code was copied. The relevant method carries an attribution comment.

## Verification (local, PHP 8.5.4 with the bcmath extension)

- **PHPUnit**: 279 tests, 1310 assertions, **OK** (added directional / high-precision / bcdivmod tests)
- **PHPStan** (level max): **No errors**
- **php-cs-fixer**: **0** findings
- **Native parity**: `bcround` checked over 8 modes × 21 inputs (including high precision and negative precision) → **0 mismatches**; `bcdivmod` matches too, and division by zero throws `DivisionByZeroError('Division by zero')`
- **Performance** (N=20000, ms): dropping the BigInteger dependency speeds up the rounding functions

  | Operation | old pkg | new pkg | native |
  |---|---|---|---|
  | `bcround(HalfAway)` | 156.5 | **31.6** | 1.9 |
  | `bcround` 30-digit high precision | 184.6 | **31.4** | 1.5 |
  | `bcceil` | 128.5 | **28.5** | 1.3 |

## PHPT skips resolved (no-extension environment)

The following **8** official phpt tests, previously skipped in `docker-phpt-tests`, now pass and were removed from `--skip`:
`bcdivmod, bcdivmod_by_zero, bcround_all, bcround_away_from_zero, bcround_ceiling, bcround_floor, bcround_toward_zero, bcround_early_return`

(The remaining `scale_ini` / `gh20006` / `bcround_precision_bounds` skips depend on the INI setting or the `BcMath\Number` OOP API and are out of scope.)

## Compatibility / impact

- All previously-passing tests still pass (HalfAway/HalfTowards behavior, negative-zero normalization, trailing-zero padding, the precision-bound `ValueError`, and the version-dependent `is_numeric` behavior are preserved).
- Default-mode (`HalfAwayFromZero`) and single-argument `floor`/`ceil` results are unchanged (no impact on EC-CUBE's usage pattern; verified in #81).

## Note on the PHPStan job

The `PHPStan` job was **already failing on `main`** with a single PHP 8.1-only false positive (`in_array()` reported as always-true at `tests/BCMathTest.php`). It is unrelated to this PR; it is silenced here so the job can go green.

Refs #81
Closes #36
Closes #39

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/BCMath.php [hunk 1]
```diff
@@ -600,6 +585,33 @@ public static function mod(string $num1, string $num2, ?int $scale = null): stri
...
+        [$num1, $num2] = self::validateAndNormalizeInputs($num1, $num2, 'bcdivmod');
+        $scale = self::resolveScale($scale);
+        self::validateScale($scale, 'bcdivmod', 3);
+        self::checkDivisionByZero($num2);
+
+        return [
+            self::div($num1, $num2, 0),
+            self::mod($num1, $num2, $scale),
+        ];
+    }
```
[the remark this answers is not in the document]
Comment (nanasess) [reply src/BCMath.php:622]: ご指摘ありがとうございます。除算を1回に統合しました（commit 6267255）。

ただし phpseclib の `divide()` が返す剰余は**非負規約**で、`bcmod()` の切り捨て規約と異なり負のオペランドで結果がずれました（例: `bcdivmod('-17','5')` が `['-3','3']` になる）。そのため剰余は `divide()` の戻り値をそのまま使わず、`$num1Big - $num2Big * $quotient` で明示計算しています。これで除算は1回のまま、ネイティブ `bcdivmod()` と全ケース一致することを確認しました。

File: src/BCMath.php [hunk 2]
```diff
@@ -1033,143 +1033,243 @@ public static function round(string $num, int $precision = 0, $mode = PHP_ROUND_
...
+    public static function bcroundHelper(string $number, int $precision, int $mode = PHP_ROUND_HALF_UP): string
+    {
+        $token = match ($mode) {
+            PHP_ROUND_HALF_DOWN => self::ROUND_HALF_TOWARDS_ZERO,
+            PHP_ROUND_HALF_EVEN => self::ROUND_HALF_EVEN,
+            PHP_ROUND_HALF_ODD => self::ROUND_HALF_ODD,
+            default => self::ROUND_HALF_AWAY_FROM_ZERO,
+        };
+
+        return self::roundDigits($number, $precision, $token);
```
[the remark this answers is not in the document]
Comment (nanasess) [reply src/BCMath.php]: ご指摘のとおりです。公開 helper が検証を迂回していたため、`round()` へ委譲して数値・precision・mode 検証を共通化しました（commit 6267255）。これにより不正な数値や巨大 precision も `round()` と同じガードを通ります。

File: src/BCMath.php [hunk 3]
```diff
@@ -1033,143 +1033,243 @@ public static function round(string $num, int $precision = 0, $mode = PHP_ROUND_
...
+            $shift = -$precision;
+            $intLength = strlen($intPart);
+            if ($shift <= $intLength) {
+                $splitPos = $intLength - $shift;
+                $scaledInt = substr($intPart, 0, $splitPos);
+                $scaledInt = $scaledInt === '' ? '0' : $scaledInt;
+                $scaledFrac = substr($intPart, $splitPos).$fracPart;
+            } else {
+                $scaledInt = '0';
+                $scaledFrac = str_repeat('0', $shift - $intLength).$intPart.$fracPart;
```
[the remark this answers is not in the document]
Comment (nanasess) [reply src/BCMath.php]: ご指摘ありがとうございます。対応しました（commit 6267255）。

- `roundDigits()`: dropped 桁は先頭桁（この分岐では常に '0'）と非ゼロ有無だけを見るため、`($shift - $intLength)` 個のゼロ確保をやめ圧縮表現（'01' / '0'）に変更
- `placeDecimalPoint()`: ゼロ結果は `str_repeat` 前に短絡

`bcround('5', -2000000000, HalfAwayFromZero)` が OOM せず '0' を返すこと、`AwayFromZero` 等で非ゼロの巨大結果になる場合はネイティブ同様に生成することを確認しました。

File: phpstan.neon.dist [hunk 4]
```diff
@@ -29,7 +29,7 @@ parameters:
 			message: "#Match arm comparison between RoundingMode and '(halfawayfromzero|halftowardszero|halfeven|halfodd|towardszero|awayfromzero|negativeinfinity|positiveinfinity)' is always false\\.#"
 		-
 			identifier: argument.type
-			message: "#Parameter \\#3 \\$mode of (static method bcmath_compat\\\\BCMath::round\\(\\)|function bcround) expects int\\|RoundingMode, (string|int) given\\.#"
+			message: "#Parameter \\#3 \\$mode of (static method bcmath_compat\\\\BCMath::round\\(\\)|function bcround) expects (int\\|RoundingMode|RoundingMode), (string|int) given\\.#"
```
[the remark this answers is not in the document]
Comment (nanasess) [reply phpstan.neon.dist:32]: こちらは現行のパターンが正しいため、変更を見送らせてください。

PHPStan が実際に出力するメッセージの「given」側は常に**単一の具体型**で、`... int given`（ネイティブ bcround シグネチャ, PHP 8.4+）または `... string given`（ポリフィルシグネチャ）のいずれかです。`string|int` という結合型がそのまま出力されることはありません。

したがって `(string|int)` の交替は「`string` または `int`」を意図どおり正しくマッチします。むしろ `string\|int`（リテラル化）にすると単一の `int given` にマッチしなくなり ignore が効かなくなります。ローカル（8.5, 拡張あり→`int given`）と CI（8.1→ポリフィル型）双方でこのパターンが機能することを確認済みです。

File: src/BCMath.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (nanasess) [reply src/BCMath.php:622]: @gemini-code-assist
